linkerd/trace-context/src/propagation/w3c.rs

```diff
@@ -0,0 +1,148 @@
+use http::HeaderValue;
+use tracing::{trace, warn};
+
+use super::{decode_id_with_padding, get_header_str, Propagation, TraceContext};
+use crate::{Flags, Id};
+
+pub const HTTP_TRACEPARENT: &str = "traceparent";
+pub const HEADER_VERSION_00: &str = "00";
+
+pub fn unpack_w3c_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
+    get_header_str(request, HTTP_TRACEPARENT).and_then(parse_context)
+}
+
+/// Given an http request and a w3c trace context, create a new Span ID and
+/// assign it to the tracecontext header value, in order to propagate
+/// the trace context downstream.
+pub fn increment_http_span_id<B>(request: &mut http::Request<B>, context: &TraceContext) -> Id {
+    let span_id = Id::new_span_id(&mut rand::thread_rng());
+
+    trace!("incremented span id: {}", span_id);
+
+    let new_header = {
+        let mut buf = String::with_capacity(60);
+        buf.push_str(HEADER_VERSION_00);
+        buf.push('-');
+        buf.push_str(&hex::encode(context.trace_id.as_ref()));
+        buf.push('-');
+        buf.push_str(&hex::encode(context.parent_id.as_ref()));
+        buf.push('-');
+        buf.push_str(&hex::encode(vec![context.flags.0]));
+        buf
+    };
+
+    if let Result::Ok(hv) = HeaderValue::from_str(&new_header) {
+        request.headers_mut().insert(HTTP_TRACEPARENT, hv);
+    } else {
+        warn!("invalid value {} for tracecontext header", new_header);
+    }
+
+    span_id
+}
+
+/// Parse a given header value as a w3c TraceContext value.
+fn parse_context(header_value: &str) -> Option<TraceContext> {
+    let rest = match header_value.split_once('-') {
+        Some((version, rest)) => {
+            if version != HEADER_VERSION_00 {
+                warn!(
+                    "Tracecontext header {} contains invalid version: {}",
+                    header_value, version
+                );
```

Comment:
we could also
```suggestion
                warn!(%header_value, %version, "Invalid tracecontext header version");
```

Reply:
Yeah I was thinking of this. I remember @hawkw gave this feedback on the validator. I'm still unsure about the preference here, do we always opt for structured traces/logs instead of formatting the args in a string?

Reply:
In general, yeah, unless we have a string message that only makes sense with values interpolated into it.